[DOC] Configure broker certificates per listener

laidan6000 opened the pull request:
### Type of change

- Documentation

### Description

This pull request adds documentation on installing certificates on a per-listener basis (for TLS and external listeners). It updates the _Kafka cluster configuration_ section and the _Security_ chapter.

It also reinstates the following procedure that was hidden (commented out) from the documentation:

"Renewing your own CA certificates"

~~It also includes a minor fix to an unrelated procedure:~~

~~"Deploying a `KafkaConnector` resource to Kafka Connect"~~

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [X] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

stanlyDoge commented on documentation/modules/security/proc-installing-certs-per-listener.adoc:
```diff
@@ -0,0 +1,83 @@
+[id='proc-installing-certs-per-listener-{context}']
+= Installing your own CA certificates for listeners
+
+You can install your own certificates and public keys for Kafka brokers on a per-listener basis. These certificates can be signed by an external Certificate Authority (CA), such as your company's internal CA or a public CA, rather than the cluster or clients CAs (which are internal to Strimzi). Installing per-listener certificates is supported for two types of listeners:
```
I am not sure about mentioning Strimzi here. Should we use `{ProductName}`?

laidan6000 replied:
Good spot - thank you!

PaulRMellor commented on documentation/modules/security/con-certificate-authorities.adoc:
```diff
@@ -26,3 +26,15 @@ You can configure the validity period of:
+
+== Options for installing your own CA certificates
+
+In your security model, client applications might need to trust certificates that were issued by an external CA, such as your company's internal CA or a public CA, in order to connect to external systems. {ProductName} provides two methods for configuring your Kafka client applications to connect to the brokers in your Kafka cluster using your own certificates.
```
Could _security model_ be misunderstood? Could we skip the phrase and start with 'Client applications...?

PaulRMellor commented on documentation/modules/security/con-certificate-authorities.adoc:
```diff
@@ -26,3 +26,15 @@ You can configure the validity period of:
+
+You can install your own certificates at two levels:
+
+* At a cluster level--see xref:installing-your-own-ca-certificates-{context}[].
```
```suggestion
* At the cluster level--see xref:installing-your-own-ca-certificates-{context}[].
```

laidan6000 replied:
Updated.

PaulRMellor commented on documentation/modules/security/con-certificate-authorities.adoc:
```diff
@@ -26,3 +26,15 @@ You can configure the validity period of:
+
+* At a cluster level--see xref:installing-your-own-ca-certificates-{context}[].
+
+* For each listener (TLS and external) in a Kafka broker--see xref:proc-installing-certs-per-listener-{context}[].
```
TLS and external or TLS or external?

laidan6000 replied:
Good spot. I've reworded as follows:

"For each listener (TLS or external listener, or both) in a Kafka broker..."

PaulRMellor commented on documentation/modules/security/con-certificate-authorities.adoc:
```diff
@@ -26,3 +26,15 @@ You can configure the validity period of:
+
+* For each listener (TLS and external) in a Kafka broker--see xref:proc-installing-certs-per-listener-{context}[].
+
+If you install your own certificates using one or both of these methods, they are not automatically renewed by the Cluster Operator. You must manually renew certificates according to your security model.
```
Couple of mentions of _security model_. Would that be understood? How might it reflect on the manual renewal? Maybe we can drop and just say you need to manually renew the certs.

scholzj replied:
I would probably just say _You must manually renew certificates when needed._ Which could be expiration or whatever security policy the user follows. I would be a bit worried that _security model_ might not have clear meaning and understanding for everyone.

laidan6000 replied:
Removed "security model" and reworded as follows:

Your client applications might need to trust certificates that were issued by an external CA, such as your company's internal CA or a public CA, in order to connect to external systems.
To support this scenario, {ProductName} provides two methods for configuring your Kafka client applications to connect to the brokers in your Kafka cluster using your own certificates.

You can install your own certificates at two levels:

* At the cluster level--see xref:installing-your-own-ca-certificates-{context}[].

* For each listener (TLS or external listener, or both) in a Kafka broker--see xref:proc-installing-certs-per-listener-{context}[].

If you install your own certificates, they are not automatically renewed by the Cluster Operator.
In this case, you must manually renew certificates when needed.

laidan6000 replied:
This was refactored again in response to Jakub's comment that the above module "mixes together" the two types of certificates.

PaulRMellor commented on documentation/modules/security/proc-installing-certs-per-listener.adoc:
```diff
@@ -0,0 +1,83 @@
+[id='proc-installing-certs-per-listener-{context}']
+= Installing your own CA certificates for listeners
+
+You can install your own certificates and public keys for Kafka brokers on a per-listener basis. These certificates can be signed by an external Certificate Authority (CA), such as your company's internal CA or a public CA, rather than the cluster or clients CAs (which are internal to Strimzi). Installing per-listener certificates is supported for two types of listeners:
```
Not sure this needs to be in brackets: (which are internal to Strimzi)

PaulRMellor commented on documentation/modules/security/proc-installing-certs-per-listener.adoc:
```diff
@@ -0,0 +1,83 @@
+
+You can install your own certificates and public keys for Kafka brokers on a per-listener basis. These certificates can be signed by an external Certificate Authority (CA), such as your company's internal CA or a public CA, rather than the cluster or clients CAs (which are internal to Strimzi). Installing per-listener certificates is supported for two types of listeners:
+
+* TLS listeners for inter-cluster communication
```
Just checking -- not inter-broker?

ppatierno commented on documentation/assemblies/assembly-kafka-broker-configuration.adoc:
```diff
@@ -14,7 +14,7 @@
 
 * Security (Encryption, Authentication, and Authorization)
-* Listener configuration
+* Listener configuration (apart from installing per-broker certificates)
```
I think this part is misleading. This section talks about broker specific config properties that you find in the official Apache Kafka documentation. There is no parameter for specifying "per-broker certificates" which is something that Strimzi adds. The sentece in parenthesis let me think that I can, in this `config` section, set "per-broker certificates" as the only listener configuration part. It's not clear that this happens in a different way and not here.

scholzj replied:
Agree with Paolo, his think this should remain unchanged - ntohing really changed to this and you cannot configure the custom certificates in the `config` section anyway.

laidan6000 replied:
Thanks for the explanation. I've removed the parentheses.

ppatierno commented on documentation/modules/kafkaconnectors/proc-deploying-kafkaconnector.adoc:
```diff
@@ -43,12 +43,12 @@ spec:
 [source,shell,subs="+quotes"]
 ----
-oc get all --selector strimzi.io/cluster: my-connect-cluster -o name
+kubectl get all --selector strimzi.io/cluster: my-connect-cluster -o name
```
why these two changes in this PR. I am for avoiding stuff not related to the PR but opening a new one.

ppatierno replied:
I would revert this change here. It's addressed by PR #2422

laidan6000 replied:
I'll avoid doing this in the future. 

I'll attempt to revert the change now that #2422 has been submitted.

ppatierno commented on documentation/modules/security/con-certificate-authorities.adoc:
```diff
@@ -6,16 +6,16 @@
+Both the cluster CA and the clients CA have a self-signed public key certificate.
 
-Kafka brokers are configured to trust certificates signed by either the clients CA or the cluster CA. Components to which clients do not need to connect, such as ZooKeeper, only trust certificates signed by the cluster CA. Client applications that perform mutual TLS authentication have to trust the certificates signed by the cluster CA.
+Kafka brokers are configured to trust certificates signed by either the clients CA or the cluster CA. Components that clients do not need to connect to, such as ZooKeeper, only trust certificates signed by the cluster CA. Client applications that perform mutual TLS authentication must trust the certificates signed by the cluster CA.
```
"Client applications that perform mutual TLS authentication must trust the certificates signed by the cluster CA.".
Mutual TLS authentication is about having the TLS encryption (so somehow trusting the server) but even using TLS client authentication. The client applications must trust the certificates signed by the cluster CA not necessarily for mutual authentication, it's enough you want TLS encryption for having the need of trusting certificates signed by the cluster CA.

laidan6000 replied:
I think I understand now. I've rephrased the information and included a link to the new concept on mutual TLS authentication.

> Unless TLS encryption for external listeners is disabled, client applications must trust certificates signed by the cluster CA. This is also true for client applications that perform [mutual TLS authentication](https://strimzi.io/docs/master/#con-mutual-tls-authentication-using-uo).#

^ Please suggest further edits if needed.

ppatierno commented on documentation/modules/security/proc-installing-certs-per-listener.adoc:
```diff
@@ -0,0 +1,83 @@
+
+* External listeners (including `route`, `loadbalancer`, and `nodeport` types) for communication between Kafka clients and Kafka brokers
+
+Installing your own certificates and keys for plain listeners is not supported.
```
I don't think this sentence is needed. Plain listeners aren't encrypted ... they are plain so TLS isn't in place, so having certificates don't make sense.

laidan6000 replied:
Ah, I see. I'll remove this sentence.

ppatierno commented on documentation/modules/security/proc-installing-certs-per-listener.adoc:
```diff
@@ -0,0 +1,83 @@
+* A Kubernetes cluster.
+* The Cluster Operator is running.
+* Public key of the external CA.  
+* Your own X.509 certificate in CRT or PEM format.
```
CRT is not a format. It's just a possible extension for a certificate formatted with PEM. Just fyi, other formats are DER, PKCS12, etc etc. In this guess it's enough saying "...in PEM format".

scholzj replied:
You should clarify that _Your own_ certificate has to be signed by the CA maybe.

scholzj commented on documentation/assemblies/security/assembly-security.adoc:
```diff
@@ -10,9 +10,11 @@ Communication between Kafka brokers (interbroker communication), between ZooKeep
+The Cluster Operator automatically sets up and renews TLS certificates to enable encryption and authentication within your cluster.
 It also sets up other TLS certificates if you want to enable encryption or TLS authentication between Kafka brokers and clients.
 
+You can install your own Kafka broker certificates on a per-listener basis.
```
I wonder if we should call it _your own Kafka listener certificates_ instead. Broker certificates might be confusing since they are not per-broker but per-listener.

laidan6000 replied:
I like _Kafka listener certificates _, as long as we explain that they are installed for each broker.

Choosing the best terminology is so important.

scholzj commented on documentation/modules/security/con-certificate-authorities.adoc:
```diff
@@ -26,3 +26,15 @@ You can configure the validity period of:
 * Cluster CA certificates in `Kafka.spec.clusterCa.validityDays`
 * Client CA certificates in `Kafka.spec.clientsCa.validityDays`
+
+== Options for installing your own CA certificates
```
I think this mixes two things which are separate concepts:
* Suplying your own CA certificates
* Suplying server certficates for the TLS and External listeners.
They should either no be mixed into the same chapter named after nly one of these or we need to find new name.

laidan6000 replied:
Thanks for the clarification. Tom thought that we needed a short concept to distinguish between these two separate but related tasks.

I'll re-word and ask you to re-review.

scholzj commented on documentation/modules/security/con-certificate-authorities.adoc:
```diff
@@ -26,3 +26,15 @@ You can configure the validity period of:
+
+You can install your own certificates at two levels:
+
+* At a cluster level--see xref:installing-your-own-ca-certificates-{context}[].
```
I'm not sure the cluster level make sense as a name for it. I would call it _Providing your own CA_ or something.

scholzj commented on documentation/modules/security/con-certificate-authorities.adoc:
```diff
@@ -26,3 +26,15 @@ You can configure the validity period of:
+
+* At a cluster level--see xref:installing-your-own-ca-certificates-{context}[].
+
+* For each listener (TLS and external) in a Kafka broker--see xref:proc-installing-certs-per-listener-{context}[].
```
I perosnally would probably leave out the `in a Kafka broker` part.

Also - and I'm not sure we need to mention it here - it applies to external listeners using TLS only. But I guess that migth be obvious and does nto need to be mentioned here, or?

scholzj commented on documentation/modules/security/proc-installing-certs-per-listener.adoc:
```diff
@@ -0,0 +1,83 @@
+// assembly-security.adoc
+
+[id='proc-installing-certs-per-listener-{context}']
+= Installing your own CA certificates for listeners
```
These would be server certificates, not CA certificates. We need to distinguish the two options properly.

scholzj commented on documentation/modules/security/proc-installing-certs-per-listener.adoc:
```diff
@@ -0,0 +1,83 @@
+[id='proc-installing-certs-per-listener-{context}']
+= Installing your own CA certificates for listeners
+
+You can install your own certificates and public keys for Kafka brokers on a per-listener basis. These certificates can be signed by an external Certificate Authority (CA), such as your company's internal CA or a public CA, rather than the cluster or clients CAs (which are internal to {ProductName}). Installing per-listener certificates is supported for two types of listeners:
```
Should we use configure instead of install? The install doesn't sound right here, I would say the operator installs them,m the user just configures them.

laidan6000 replied:
Sounds good to me.

scholzj commented on documentation/modules/security/proc-installing-certs-per-listener.adoc:
```diff
@@ -0,0 +1,83 @@
+[id='proc-installing-certs-per-listener-{context}']
+= Installing your own CA certificates for listeners
+
+You can install your own certificates and public keys for Kafka brokers on a per-listener basis. These certificates can be signed by an external Certificate Authority (CA), such as your company's internal CA or a public CA, rather than the cluster or clients CAs (which are internal to {ProductName}). Installing per-listener certificates is supported for two types of listeners:
```
Thsi would be much easier to comment on if you followed the one sentence per line policy. We should mention somewhere here that the certificates should be either wildcard certificates with appropriatte subject or have SANs matching all possibl broker names. We should possibly include some examples of how the addresses would look like so that the user knows what should be used.

laidan6000 replied:
OK. I'll use one sentence per line as much as possible.

@stanlyDoge - Can you provide some example addresses in the correct format?

laidan6000 replied:
I've added a new module named "SAN certificates for Kafka listeners" containing the example addresses.

scholzj commented on documentation/modules/security/proc-installing-certs-per-listener.adoc:
```diff
@@ -0,0 +1,83 @@
+
+* TLS listeners for inter-cluster communication
+
+* External listeners (including `route`, `loadbalancer`, and `nodeport` types) for communication between Kafka clients and Kafka brokers
```
Maybe we can add here that the external listeners only as long as they use TLS. It is also missing `ingress` type.

scholzj commented on documentation/modules/security/proc-installing-certs-per-listener.adoc:
```diff
@@ -0,0 +1,83 @@
++
+[source,shell,subs="+quotes"]
+----
+kubectl create secret generic _my-secret_ --from-file=_my-listener-key.key_ --from-file=_my-listener-certificate.crt_
```
So do we need to clarify how the CA public key should be used? You listed it as prerequisite so it probably should be explained.

laidan6000 replied:
Can you give me a few pointers here, @stanlyDoge ?

laidan6000 replied:
I removed the prerequisite and added the following text:

> Your client applications should use the CA public key as a trusted certificate in order to verify the identity of the Kafka broker.

scholzj commented on documentation/modules/security/proc-installing-your-own-ca-certificates.adoc:
```diff
@@ -3,9 +3,9 @@
+= Installing your own CA certificates for a Kafka cluster
 
-This procedure describes how to install your own CA certificates and private keys instead of using CA certificates and private keys generated by the Cluster Operator.
+This procedure describes how to install your own CA certificates and private keys at the cluster level instead of using CA certificates and private keys generated by the Cluster Operator.
```
Again, I would probably leave out the cluster level. I do not think it is understandable what it means but people not really familiar with Strimzi.

laidan6000 commented:
This is now ready for re-review by @scholzj 

@ppatierno - please can you check that I've addressed your comment mutual TLS authentication and approve the PR if you're happy.

ppatierno commented on documentation/modules/security/con-certificate-authorities.adoc:
```diff
@@ -6,20 +6,28 @@
 
-Kafka brokers are configured to trust certificates signed by either the clients CA or the cluster CA. Components to which clients do not need to connect, such as ZooKeeper, only trust certificates signed by the cluster CA. Client applications that perform mutual TLS authentication have to trust the certificates signed by the cluster CA.
+Kafka brokers are configured to trust CA certificates signed by either the cluster CA or clients CA.
+Components that clients do not need to connect to, such as ZooKeeper, only trust CA certificates signed by the cluster CA.
```
why we added "CA" ... so starting to use "CA certificates" instead of "certificates". A client should trust a broker certificates (which is not a CA certificate) signed by the cluster CA.

laidan6000 replied:
I've ended up confused myself with the terminology :(

I'll remove the "CA".

ppatierno commented on documentation/modules/security/con-certificate-authorities.adoc:
```diff
@@ -6,20 +6,28 @@
-Kafka brokers are configured to trust certificates signed by either the clients CA or the cluster CA. Components to which clients do not need to connect, such as ZooKeeper, only trust certificates signed by the cluster CA. Client applications that perform mutual TLS authentication have to trust the certificates signed by the cluster CA.
+Kafka brokers are configured to trust CA certificates signed by either the cluster CA or clients CA.
+Components that clients do not need to connect to, such as ZooKeeper, only trust CA certificates signed by the cluster CA.
+Unless TLS encryption for external listeners is disabled, client applications must trust CA certificates signed by the cluster CA.
```
I am fine with this one but still this usage of "CA certificates".

laidan6000 replied:
Removed.

scholzj commented on documentation/assemblies/security/assembly-security.adoc:
```diff
@@ -10,9 +10,12 @@ Communication between Kafka brokers (interbroker communication), between ZooKeep
 For the Kafka and {ProductName} components, TLS certificates are also used for authentication.
 
-The Cluster Operator automatically sets up TLS certificates to enable encryption and authentication within your cluster.
+The Cluster Operator automatically sets up and renews TLS certificates to enable encryption and authentication within your cluster.
```
Do we need to clarify that it renews only its own certificates? Not any supplied by the user?

scholzj commented on documentation/modules/security/con-kafka-listener-certificates.adoc:
```diff
@@ -0,0 +1,15 @@
+
+= Kafka listener certificates
+
+In order to connect to external systems such as {ProductName}, your client applications might need to trust server certificates that were issued by an external CA, such as your company's private CA or a public CA.
```
I do not think this sentence makes sense. I think it should say something like this:
_Sometimes, you might want to use {ProductName} with your own CA such as  your company's private CA or a public CA._
or maybe:
_Sometimes you want to leverage your existing security infrastructure such as your company's private CA or a public CA._

laidan6000 replied:
Very nice. I feel that the verb _leverage_ demands an action. 

> Sometimes you might want to leverage your existing security infrastructure, such as your company's private CA or a public CA, when integrating your client applications with {ProductName}. 

I'll remove the whole sentence beginning with "In order to...".

scholzj commented on documentation/modules/security/ref-formats-kafka-listener-certificates.adoc:
```diff
@@ -0,0 +1,77 @@
+
+[id='ref-formats-kafka-listener-certificates-{context}']
+
+= SAN certificates for Kafka listeners
```
```suggestion
= Alternative subjects in server certificates for Kafka listeners
```

scholzj commented on documentation/modules/security/ref-formats-kafka-listener-certificates.adoc:
```diff
@@ -0,0 +1,77 @@
+
+= SAN certificates for Kafka listeners
+
+If you provide your own xref:kafka-listener-certificates-{context}[Kafka listener certificates], you must use a Subject Alternative Name (SAN) certificate for each listener. The SAN certificate must specify hostnames for:
```
Sorry if I confused you about this. The SANs are a feature/configuration  of the certificate. It is not certificate type. So I think it should sound something like this:

```suggestion
In order to use TLS hostname verification with your own xref:kafka-listener-certificates-{context}[Kafka listener certificates], you must use the correct Subject Alternative Names (SAN) for each listener. The certificate SANs must specify hostnames for:
```

scholzj commented on documentation/modules/security/ref-formats-kafka-listener-certificates.adoc:
```diff
@@ -0,0 +1,77 @@
+
+* The Kafka cluster bootstrap service
+
+The Kafka broker hostnames can contain wildcard names.
```
```suggestion
You can use wildcsard certificates if supported by your CA.
```

scholzj commented on documentation/modules/security/ref-formats-kafka-listener-certificates.adoc:
```diff
@@ -0,0 +1,77 @@
+
+The Kafka broker hostnames can contain wildcard names.
+
+Using SAN certificates is required for TLS hostname verification.
```
```suggestion
Using certificates with the coorrect SANs is required for TLS hostname verification.
```

laidan6000 replied:
I'll remove this sentence because we now mention TLS hostname verification right at the start.

scholzj commented on documentation/modules/security/ref-formats-kafka-listener-certificates.adoc:
```diff
@@ -0,0 +1,77 @@
+
+Using SAN certificates is required for TLS hostname verification.
+
+== TLS listener SAN certificate examples
```
```suggestion
== TLS listener SAN examples
```

scholzj commented on documentation/modules/security/ref-formats-kafka-listener-certificates.adoc:
```diff
@@ -0,0 +1,77 @@
+
+== TLS listener SAN certificate examples
+
+Use the following examples to help you specify hostnames in SAN certificates for TLS listeners.
```
```suggestion
Use the following examples to help you specify hostnames the SANs in your certificates for TLS listeners.
```

scholzj commented on documentation/modules/security/ref-formats-kafka-listener-certificates.adoc:
```diff
@@ -0,0 +1,77 @@
+_<cluster-name>_-kafka-bootstrap._<namespace>_.svc
+----
+
+== External listener SAN certificate examples
```
```suggestion
== External listener SAN examples
```

scholzj commented on documentation/modules/security/ref-formats-kafka-listener-certificates.adoc:
```diff
@@ -0,0 +1,77 @@
+
+== External listener SAN certificate examples
+
+For external listeners which have TLS encryption enabled, the hostnames you need to specify in SAN certificates depends on the external listener `type`.
```
```suggestion
For external listeners which have TLS encryption enabled, the hostnames you need to specify in certificates depends on the external listener `type`.
```

scholzj commented on documentation/modules/security/ref-formats-kafka-listener-certificates.adoc:
```diff
@@ -0,0 +1,77 @@
+|===
+
+¦External listener type
+¦In the SAN certificate, specify...
```
```suggestion
¦In the SANs, specify...
```

scholzj commented on documentation/modules/security/ref-formats-kafka-listener-certificates.adoc:
```diff
@@ -0,0 +1,77 @@
+¦In the SAN certificate, specify...
+
+m¦Route
+¦Addresses of all Kafka broker `Routes` and the address of the bootstrap service `Route`.
```
```suggestion
¦Addresses of all Kafka broker `Routes` and the address of the bootstrap `Route`.
```

scholzj commented on documentation/modules/security/ref-formats-kafka-listener-certificates.adoc:
```diff
@@ -0,0 +1,77 @@
+You can use a matching wildcard name.
+
+m¦loadbalancer
+¦Addresses of all Kafka broker `loadbalancers` and the `loadbalancer` bootstrap address.
```
```suggestion
¦Addresses of all Kafka broker `loadbalancers` and the bootstrap `loadbalancer` address.
```

laidan6000 commented:
Thanks - I've made those changes to the terminology for SANs, etc. I didn't use your suggested changes to avoid running into any problems with the DCO check - I made the changes manually instead.

laidan6000 commented:
Reorganised the content slightly to reflect minimalism principles. The concept now clearly defines Kafka listener certificates rather than leaving the definition until the procedure. No substantive changes to the technical content were made.

scholzj commented on documentation/modules/security/con-kafka-listener-certificates.adoc:
```diff
@@ -6,10 +6,15 @@
-Sometimes you might want to leverage your existing security infrastructure, such as your company's private CA or a public CA, when integrating your client applications with {ProductName}. 
-To support this requirement, {ProductName} allows you to provide your own _Kafka listener certificates_ and public keys for external listeners which have TLS encryption enabled.
-Your Kafka client applications will connect to Kafka brokers using Kafka listener certificates rather than CA certificates signed by the cluster CA or clients CA.
+You can provide your own server certificates and CA public keys for the following types of listeners:
```
Sorry, I had to get back to this. I do not think the CA belongs here.

```suggestion
You can provide your own server certificates and public keys for the following types of listeners:
```

laidan6000 replied:
OK, that's updated now.

scholzj commented:
Thanks @laidan6000
